storage/filesystem.go

```diff
@@ -176,18 +189,54 @@ func (f *Filesystem) RegisterMetadata(meta metadata.Metadata) (bool, error) {
 	return false, nil
 }
 
-func (f *Filesystem) WriteFiles(volumeID string, files map[string][]byte) error {
-	if err := os.MkdirAll(f.dataPathForVolumeID(volumeID), 0644); err != nil {
+// WriteFiles writes the given data to filesystem files within the volume's
+// data directory. Filesystem supports changing ownership of the data directory
+// to a custom gid.
+func (f *Filesystem) WriteFiles(meta metadata.Metadata, files map[string][]byte) error {
+	// Data directory should be read, write and execute only to the fs user; read and executable to group
+	if err := os.MkdirAll(f.dataPathForVolumeID(meta.VolumeID), 0550); err != nil {
+		return err
+	}
+
+	fsGroup, err := f.fsGroupForMetadata(meta)
+	if err != nil {
 		return err
 	}
 
-	writer, err := util.NewAtomicWriter(f.dataPathForVolumeID(volumeID), fmt.Sprintf("volumeID %v", volumeID))
+	// If a fsGroup is defined, Chown the directory to that group.
+	if fsGroup != nil {
+		if err := os.Chown(f.dataPathForVolumeID(meta.VolumeID), -1, int(*fsGroup)); err != nil {
+			return fmt.Errorf("failed to chown data dir to gid %v: %w", *fsGroup, err)
+		}
+	}
+
+	writer, err := util.NewAtomicWriter(f.dataPathForVolumeID(meta.VolumeID), fmt.Sprintf("volumeID %v", meta.VolumeID))
 	if err != nil {
 		return err
 	}
 
 	payload := makePayload(files)
-	return writer.Write(payload)
+	if err := writer.Write(payload); err != nil {
+		return err
+	}
+
+	// If a fsGroup is defined, Chown all files within the data directory.
+	if fsGroup != nil {
+		dirName := f.dataPathForVolumeID(meta.VolumeID)
+		entries, err := os.ReadDir(dirName)
```

Comment:
Given the atomic writer creates subdirectories containing the actual files, does this work..? It seems like this would chown the directories containing data files, rather than the data itself?

Reply:
These are the files that get `Chowned` here:

```
csi-data-dir/inmemfs/csi-c3d3286f77ef69343fa88e568da8db465d919a403ba1daa27f628597b80533a9/data/..2021_06_25_15_49_43.932262773
csi-data-dir/inmemfs/csi-c3d3286f77ef69343fa88e568da8db465d919a403ba1daa27f628597b80533a9/data/..data
csi-data-dir/inmemfs/csi-c3d3286f77ef69343fa88e568da8db465d919a403ba1daa27f628597b80533a9/data/ca.crt
csi-data-dir/inmemfs/csi-c3d3286f77ef69343fa88e568da8db465d919a403ba1daa27f628597b80533a9/data/tls.crt
csi-data-dir/inmemfs/csi-c3d3286f77ef69343fa88e568da8db465d919a403ba1daa27f628597b80533a9/data/tls.key
```

Is this not correct/what we want?

Reply:
I have changed this to now Chmoding the real underlying files. We construct the path by using the atomic_writer's "..data" dir.

```go
	// If a fsGroup is defined, Chown all files just written.
	if fsGroup != nil {
		// "..data" is the well-known location where the atomic writer links to the
		// latest directory containing the files just written. Chown these real
		// files.
		dirName := filepath.Join(f.dataPathForVolumeID(meta.VolumeID), "..data")

		for filename := range files {
			// Set the uid to -1 which means don't change ownership in Go.
			if err := os.Chown(filepath.Join(dirName, filename), -1, int(*fsGroup)); err != nil {
				return err
			}
		}
	}
```